### ISSUE TITLE
Create CombinedCodeView (MemoryView/DisassemblyView+CodeView)

### ISSUE THREAD

**Issue by ptomin** [uxmal/reko#155]
It would be nice to view both memory and code in one tab. It allows user to compare memory/disassembly with decompiled code. I suggest to create CombinedCodeView for it
Following screen shots contain suggested implementation of CombinedCodeView 

![sample1](https://cloud.githubusercontent.com/assets/13188041/13254594/c9285caa-da63-11e5-8542-c6dba566f7bb.jpg)
![sample2](https://cloud.githubusercontent.com/assets/13188041/13254602/d02924b2-da63-11e5-894b-c1603230ec10.jpg)

I have crude implementation of it: adefe5a

**Comment by halsten** [comment]
I really like this idea.

**Comment by uxmal** [comment]
The previous item #123 also takes this up. I've been wanting this for a long time. This solution is is great prototype, and thanks for that; I particularly like the interleaving of code and data. However, it will quickly run Windows out of resources by creating all those MemoryControls and DisassemblerControls. After a few throusand of those, Windows will crawl on its knees.

What is needed is a hybrid TextView model, where memory and code could be interleaved and displayed in a single TextView by creating a two-tiered text view model. I will code a prototype tonight, and see what you think.

Still mulling over what the right side should look like. The problem is that many procedures are not contiguous in memory. I have several samples where there are jumps all over the place.

**Comment by uxmal** [comment]
First point of action is to convert the `MemoryControl` so that it renders itself with `TextSpan`s.

**Comment by uxmal** [comment]
The nested-textView branch now has a mixed mode viewer. It's still buggy but should be a good starting point.

**Comment by ptomin** [comment]
I have implemented prototype of MixedView+CodeView panel (see c4797b8) using MixedCodeDataModel and NestedTextModel. When right text view scrolled left text view moved to corresponding function/data and vice versa.
New screen shot: 
![sample1](https://cloud.githubusercontent.com/assets/13188041/13554629/d67becea-e3ce-11e5-9419-d1ea85898cf3.jpg)

**Comment by uxmal** [comment]
That's looking nice, Pavel! Obviously needs cleanup but certainly a step in the right direction. I will review your code shortly.

**Comment by uxmal** [comment]
I merged your changes. There is still a crash in there that needs to be taken care of -- do you think you could look at it? I will try to get a repro case today.

**Comment by ptomin** [comment]
I'll try to fix it

**Comment by uxmal** [comment]
I'm adding another piece to this puzzle. I just committed a few changes into the master branch
- `ByteMapView`, which shows all the loaded bytes as a grayscale bitmap. Eventually, we want to be able to click on a space in the map and have Reko navigate to the corresponding address. However, this is probably not quite as interesting as when viewing code. See screenshot.
![ssh](https://cloud.githubusercontent.com/assets/10154465/13635382/a25b0368-e5fb-11e5-8654-6f8d90ec7ced.png)
- I've incorporated Microsoft Research's AutomaticGraphLayout Graph Viewer into Reko! (do I hear @halsten cheering? :) ) It allows you to view procedures as a graph. See screen shot.

![ssi](https://cloud.githubusercontent.com/assets/10154465/13635381/a2580ece-e5fb-11e5-99a4-da19d6d54ae7.png)

It occurred to me that a user might want to switch between looking at the graph, looking at the code, and looking at the mixed  code / assembler. What do you guys think?

**Comment by ptomin** [comment]
Tasks should be done before release of CombinedCodeView
- [ ]  Create procedure or global variable declaration text box
- [ ]  Improve alignment of memory area of MixedCodeDataView
- [ ]  Implement mouse click event on procedures and variables links
- [ ]  Fix implementation of Back/Forward actions
- [ ]  Code is too close to border of CodeView(right panel).

**Comment by ptomin** [comment]
I have finished implementation of procedure or global variable declaration text box. There are 3 cases
1. Show declaration box at procedure body. It shows selected procedure signature (or procedure name if signature is not set). It is not allowed to enter variable declaration.
2. Show declaration box at global variable.  It shows selected global variable declaration. User could enter both procedure or variable declaration. If procedure declaration(or  procedure name) is entered, selected global variable will be removed and new procedure will be added
3. Show declaration box at undefined memory area. It shows empty line. User could enter both procedure or variable declaration. And procedure or variable will be added to address specified by anchor position
John, could you test it?

**Comment by uxmal** [comment]
This looks really good Pavel. In particular, the combined view has become much more stable (fewer crashes).

I'm still working on the padding feature, hope to finish it this week. Perhaps you can get Back/Forward to work?

**Comment by ptomin** [comment]
I'm going to start implementation of mouse click event on procedures and variables links

**Comment by ptomin** [comment]
I have implemented mouse click event on procedures and variables links

**Comment by ptomin** [comment]
I have modified CombinedCodeView and other classes so that it's possible to view globals when mouse pressed on "Global variables" node in project tree. And now "Global variables" node is not shown if there is no image map items except procedures and unknown types.

So one more CombinedCodeView task could be checked

**Comment by ptomin** [comment]
John you commit 30ccedc contains fix for 6th task ("Reko fails when MixedCodeDataView is scrolled to the end on Python sample"). One more task should be checked.

**Comment by uxmal** [comment]
Yes,  I haven't seen that many crashes recently.

**Comment by ptomin** [comment]
I have fixed implementation of Back/Forward actions (34b4449). Corresponding item could be checked

**Comment by uxmal** [comment]
I've checked off a few more items, and added a new one (Ctrl+C / Edit Copy should work)

**Comment by ptomin** [comment]
> (Ctrl+C / Edit Copy should work)

I have already implemented it in b41e6ec 3 weeks ago. Is it work incorrectly?

**Comment by uxmal** [comment]
Copy in the MixedCodeView drops the last character. I.e. if I select `12` and copy, I get only `1` in the clipboard.

**Comment by ptomin** [comment]
> Copy in the MixedCodeView drops the last character

Yes, it should be fixed

**Comment by ptomin** [comment]
Old CodeView contains Declaration/Characteristics panel:
![sample1](https://cloud.githubusercontent.com/assets/13188041/14434872/9cf9132c-002e-11e6-9336-8dc7b5139e05.jpg)
CodeView will go away soon. Should we use this panel in new CombinedCodeView?
New DeclarationTextBox have functionality of  Declaration tab. So the rest unimplemented tab is Characteristics ( with one checkbox)
There is other procedure edit dialog:
![sample2](https://cloud.githubusercontent.com/assets/13188041/14435161/e35d604c-002f-11e6-87ab-13348e7cbc45.jpg)
Maybe it should be used and there is necessity to move Declaration/Characteristics panel functionality to new user interface.

**Comment by uxmal** [comment]
Sounds reasonable. I liked the idea of having the "property sheet" always visible, but it does take up a lot of real estate on the screen.

**Comment by ptomin** [comment]
So what? We won't do nothing with it, is not it?

**Comment by uxmal** [comment]
It's not hard to bring it back. Let's use the Edit Properties dialog. The  "Parameters" tab is obsolete, however, and should be removed.

**Comment by ptomin** [comment]
And It seems to ignore characteristics check boxes and name editor

**Comment by uxmal** [comment]
It should be possible to transplant the interactor code behind the corresponding controls in the CodeViewPane to the dialog.

**Comment by ptomin** [comment]
> It should be possible to transplant the interactor code behind the corresponding controls in the CodeViewPane to the dialog.

1. CodeViewPane implementation does not contain interaction with "Procedure never returns" checkbox so it could not moved to the "Edit Procedure" dialog
2. Do you mean replace existing SignatureParser in "Edit Procedure" dialog to CParser?

**Comment by uxmal** [comment]
1. We'll need to add logic to listen to the "Procedure never returns" and set/clear the boolean flag `ProcedureCharacteristics.Terminates`. 
2. Yes, it should use CParser; it makes no sense to expose the internal Reko representation of data types to a user.

**Comment by ptomin** [comment]
I have fixed Edit Copy bugs. Corresponding item should be checked. Is not it, John?

**Comment by uxmal** [comment]
Indeed! I'm going to see today if I can reproduce the font error you have been experiencing. Once items on list are done it's time to prepare for a release of 0.6.0.0

**Comment by uxmal** [comment]
OK, I have made a few attempts to reproduce the font error without success. In particular, I've removed the HKCU\Software\jklSoft\Reko key in the registry, but that made no difference. Are you still seeing this error?

**Comment by ptomin** [comment]
> Are you still seeing this error?

Yes, I am. 
I'll try to debug code when I'll have spare time

**Comment by ptomin** [comment]
Interesting... UiPreferencesService::ConvertFrom does not accept "Lucida Console, 9pt" but it accepts "Lucida Console; 9pt". Maybe the reason is different locales

**Comment by uxmal** [comment]
Good catch; I have set US locale on my machine, otherwise I would likely had encountered this too. Looking into fixing this.

**Comment by uxmal** [comment]
I just checked in 4961ed74, does it fix the problem in your locale?

**Comment by ptomin** [comment]
Yes, it fixed it. Thanks!

**Comment by ptomin** [comment]
.. but ther is other problem. When I change memory/disassembly font in "User Preference" dialog padding between image map items disappears

**Comment by uxmal** [comment]
Commit 4625ea3576276781e82d7f16652fc2b062c22b06 should have fixed the missing padding problem,

**Comment by ptomin** [comment]
> Commit 4625ea3 should have fixed the missing padding problem

It works, thanks.

**Comment by ptomin** [comment]
I have modified "Edit Procedure" dialog
CParser is used to parse signatures
The state of checkboxes saves when OK clicked
![sample1](https://cloud.githubusercontent.com/assets/13188041/14532365/3c8ca33e-027a-11e6-9cb0-a9a413d484ba.jpg)

**Comment by ptomin** [comment]
All items is checked. Now it's time to push it to master

**Comment by uxmal** [comment]
I will make a release today.

**Comment by uxmal** [comment]
....and we're code complete; what a great feeling. @ptomin : tanks again for your great contributions to this feature!

**Comment by ptomin** [comment]
Great! Thank you too, John!
